### DISCUSSION TITLE
absolute area fraction by variant

### DISCUSSION THREAD

**Discussion by trudator** [ORTools4MTEX/ORTools#19] [Q&A] [answered]
Hi Frank,

Is there any way to plot the absolute fraction of each variant in the whole grain graph? To be more specific, is that possible to plot it using **EBSD data** instead of grain data? Thanks Frank!

Linh

**Comment by frankNiessen** [comment, marked as answer]
Hi, 

this might not be the most elegant way of doing it, but you can compute the variant ID of each EBSD pixel this way:

```
    pGrains = job.grains(job.mergeId(job.ebsdPrior(job.csChild).grainId));
    isParent = pGrains.phaseId == job.parentPhaseId;
    pGrains = pGrains(isParent);
    cEBSD = job.ebsdPrior(job.csChild);
    cEBSD = cEBSD(isParent);
    varIds = calcVariantId(pGrains.meanOrientation,cEBSD.orientations,job.p2c,'variantMap',job.variantMap);
```
Then you can check how often each variant ID is present within `varIds` which should give you the relative area fraction based on EBSD data. 

Please let me know if it works. 

Best wishes
Frank

**Reply by trudator** [reply]
Hi Frank,

It works! Thank you so much. However, I see a huge difference between plotting by EBSD and grain data (please refer to the below image). If I used grain data, there is variant selection at variant 1, however, it disappears when I changed to using EBSD data. To your opinion. which method is more reliable? Thanks again, Frank!

![image](https://user-images.githubusercontent.com/117386810/203853794-7359d469-d6ed-47f5-a6d5-822ae2563989.png)

**Reply by frankNiessen** [reply]
How to you determine the grain data histogram? Do you count the number of grains of each variant or do you take each grain weighted by its area? The latter should ideally be identical to EBSD, the prior will potentially give you a quite distorted result. I would either go for the EBSD data (Assuming you have cleaned it properly before reconstruction) or grain data weighted by the grain area.

**Reply by trudator** [reply]
Hi Frank,

I take the weighted area of each grain like you said (below is my code). It's quite interesting to see the difference between them. `(Assuming you have cleaned it properly before reconstruction)` Did you mean cleaning step like halfQuadraticFilter?

```
areas = job.transformedGrains.area;
vIds = job.variantId(job.isTransformed);
for ii = 1:24
      vArea(ii) = sum(areas(vIds == ii));
end

v_abs_area = vArea / sum(vArea)*100
```

**Reply by frankNiessen** [reply]
Good you take the area into account. To be honest, your histograms do look similar apart from a shift from variant 1 to variant 4. Variant 1 and 4 are within the same packet and have a very slight misorientation angle to each other. It is therefore possible that during grain reconstruction you merge some of these variants into the same grain and the distinction between variant 1 and 4 gets lost. 

Cleaning can of course be many things, I was refering to removal of very small grains that are associated with noise or local misindexing. Now I have not seen your dataset and can therefore not judge how much this is required in your case. 

One other thing to keep in mind is that the grains also cover areas that are not indexed, whereas EBSD pixels only concern the indexed data. In case you have a lot of unindexed data it might therefore be possible that you get different statistics when considering the grain areas. 

As a last comment: What do you actually want to achieve with your analysis? If you want to analyze your martensitic structure it might be more worthwhile to look at the equivalent variants of variants 1 to 6. Variants 7 to 24 are just repetitions of the same variants within different packets, but share the same characteristics as variants 1 to 6. The equivalent variants of variant 1 for example are 1, 7, 13 and 19.
We have implemented some analysis in this direction that is based on the work by Morito et al. 

<div class="csl-entry">Morito, S., Pham, A. H., Hayashi, T., &#38; Ohba, T. (2015). Block boundary analyses to identify martensite and bainite. <i>Materials Today: Proceedings</i>, <i>2</i>, S913–S916. https://doi.org/10.1016/j.matpr.2015.07.430</div>

see the function plotMap_variantPairs (with an example in ORTools_Example 7).

I hope this is helpful
Best of Luck
Frank

**Reply by trudator** [reply]
Thanks Frank. Really interesting feature. I will definitely try it shortly. A follow up question in the interest of your response above. My cleaning routine is to: remove pixels with a low confidence interval < 0.1 (I used TSL OIM) --> remove small grain after reconstructing --> fill up/denoise the reconstructed microstructure by using `halfQuadraticFilter`. Then I tried to plot the variant map by using EBSD but still ended up with a lot of unindexed pixels even though the new (denoised) microstructure was fulfilled after the last step. I doubt something wrong with my code and that's probably the reason why my ebsd-based variant distribution deviates from the grain-based one. Thanks Frank again for helping me out!

```
% remove bad CI pixel
ebsd = ebsd(ebsd.ci>0.1);
job = setParentGrainReconstructor(ebsd,grains,Ini.ebsdPath)
job.p2c = orientation.KurdjumovSachs(job.csParent, job.csChild)
job.calcParent2Child;
job.calcGraph('threshold',2.5*degree,'tolerance',2.5*degree)
job.clusterGraph('inflationPower',1.6)
plotMap_clusters(job,'linewidth',2);
job.calcParentFromGraph
job.revert(job.grains.fit > 5*degree | job.grains.clusterSize < 5)

for k = 1:5 
  job.calcGBVotes('p2c','threshold',k*2.5*degree)
  job.calcParentFromVote
end

job.mergeSimilar('threshold',15*degree);
job.mergeInclusions('maxSize',50);

% Calculate the variant IDs of all alpha grains.
job.calcVariants

[parentGrains, parentEBSD.grainId] = calcGrains(parentEBSD('indexed'),'angle',15*degree);
% remove very small grains
parentEBSD(parentGrains(parentGrains.grainSize<10)) = [];
% redo grain reconstrucion
[parentGrains, parentEBSD.grainId] = calcGrains(parentEBSD('indexed'),'angle',15*degree);
parentGrains = smooth(parentGrains,10);
% denoise
F = halfQuadraticFilter;
parentEBSD = smooth(parentEBSD('indexed'),F,'fill',parentGrains);

% plot variant map
plotMap_variants(job,'linewidth',3);
```

You can see the variant map with a lot of unindexed pixels (in white).
![image](https://user-images.githubusercontent.com/117386810/203957422-1896fea9-7197-4261-88e8-8559465595c0.png)

**Reply by frankNiessen** [reply]
Hi again, 

yes, there is something wrong. 
You load in EBSD data, then you compute grains from that data, then you remove EBSD with low CI, and finally you pass both EBSD and grains into the parentGrainReconstructor. Now your grains structure refers to EBSD data that does not exist anymore, as you have edited the EBSD data after grain reconstruction. So after filtering your EBSD data, you need to compute the grains again before initializing the parent grain reconstructor:

```
ebsd = ebsd(ebsd.ci>0.1);`
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
job = setParentGrainReconstructor(ebsd,grains,Ini.ebsdPath)
```

Best wishes
Frank

**Reply by trudator** [reply]
Hi Frank, 
Sorry for the confusion but I actually had that line in my code. I'm attaching my code here again (and still returning with the above result). Could you please check it again? Thanks for following up with me.

```
% remove bad CI pixel
ebsd = ebsd(ebsd.ci>0.1);

% Grains are calculated with a 15° threshold
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',15*degree);
% EBSD data in small grains are removed
ebsd(grains(grains.grainSize < 2)) = [];
% Recalculate the grains from the remaining data ...
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',15*degree);

job = setParentGrainReconstructor(ebsd,grains,Ini.ebsdPath)
job.p2c = orientation.KurdjumovSachs(job.csParent, job.csChild)
job.calcParent2Child;
job.calcGraph('threshold',2.5*degree,'tolerance',2.5*degree)
job.clusterGraph('inflationPower',1.6)
plotMap_clusters(job,'linewidth',2);
job.calcParentFromGraph
job.revert(job.grains.fit > 5*degree | job.grains.clusterSize < 5)

for k = 1:5 
  job.calcGBVotes('p2c','threshold',k*2.5*degree)
  job.calcParentFromVote
end

job.mergeSimilar('threshold',15*degree);
job.mergeInclusions('maxSize',50);

%% Plot reconstructed parent EBSD orientations
parentEBSD = job.calcParentEBSD;

[parentGrains, parentEBSD.grainId] = calcGrains(parentEBSD('indexed'),'angle',15*degree);
% remove very small grains
parentEBSD(parentGrains(parentGrains.grainSize<10)) = [];
% redo grain reconstrucion
[parentGrains, parentEBSD.grainId] = calcGrains(parentEBSD('indexed'),'angle',15*degree);
parentGrains = smooth(parentGrains,10);
% denoise
F = halfQuadraticFilter;
parentEBSD = smooth(parentEBSD('indexed'),F,'fill',parentGrains);

% Calculate the variant IDs of all alpha grains.
job.calcVariants

% plot variant map
plotMap_variants(job,'linewidth',3);
```

**Reply by trudator** [reply]
It's getting quite bad in some of my dataset. 

![image](https://user-images.githubusercontent.com/117386810/204028942-077db00f-fb0c-4a95-a429-975a0efcbb18.png)

**Reply by frankNiessen** [reply]
To me this seems like this is just the data you have discarded by setting the IQ<0.1 criterion. Could you plot the IPF map of the same region before and after having cleaned the data with the IQ criterion?

**Reply by trudator** [reply]
Hi Frank,

I think it's the same for the dataset used in example 1 too. As I said, even though I have used a couple of cleaning methods to smoothen/denoise my ebsd data (left image), the variant map (right image) still returns with many white pixels. My question is how to achieve plot the variant map based on ebsd data after smoothening/denoising. I have attached my code [here](https://drive.google.com/file/d/1EI5XshyqR-yatqkur7r5uPHAeKcGo_pS/view?usp=sharing) for better use. 

![image](https://user-images.githubusercontent.com/117386810/204051969-c9c21842-9f35-4012-a87c-b3b8d3f80b54.png)

I wanted to point it out since it would affect my area fraction vs. variant/packet graph in some ways (not too much in example 1, but could be a huge impact on another dataset, like mine). Here below is the two graphs from ebsd and grain data from example 1. 

![image](https://user-images.githubusercontent.com/117386810/204053207-2d954d0b-3ae3-4064-b407-396b98a8117d.png)

Thanks Frank!

**Reply by frankNiessen** [reply]
I think you misunderstand the way EBSD data and grains work in MTEX. 

This is the EBSD dataset from Example 1, it contains several unindexed points.
![image](https://user-images.githubusercontent.com/41304940/204054218-046c3496-75a1-4f6e-a4eb-2b93fc2a1181.png)

Then we compute the grains, this operation consumes all the unindexed points into the grains to reduce the effect of fragmentation
![image](https://user-images.githubusercontent.com/41304940/204054309-7195f529-8d9c-4c9c-a13c-5e3bfdb16445.png)

The map you show on the left handside is a reconstructed EBSD map in which the unindexed points have been filled with a half-quadratic filter, it is not the original EBSD dataset. 
![image](https://user-images.githubusercontent.com/41304940/204054808-f13a48bd-759a-495d-8990-36a63adb5d5b.png)

This operation does not extrapolate the variant IDs as well, so you cannot get a variant map from such postprocessed maps. 

The variant map function works either on the ebsd dataset 

`plotMap_variants(job,'linewidth',3);`

or on the grains dataset

`plotMap_variants(job,'grains','linewidth',3);`

but not on further postprocessed EBSD data.

**Reply by trudator** [reply]
I see. Thanks for your interpretation, Frank. So way back to my previous question: in case I have some huge deviation between EBSD- and grain-based area fraction vs. variant chart (either remove noise or local unindexed pixels or not), which one you would recommend using (for the sake of data transparency)? Thanks, Frank.

Here is your recommendation above. 
`I would either go for the EBSD data (Assuming you have cleaned it properly before reconstruction) or grain data weighted by the grain area.`

**Reply by frankNiessen** [reply]
It really depends on the case and there is no wrong and right as long as you clearly state what you have done in your  methods. I would probably trust the cleaned (and not smoothed) EBSD data the most, so removal of CI<0.1 and of small grains. If you don't expect a different tendency of the variants to be succesfully indexed this should give you the right picture. 

Best wishes
Frank

**Reply by trudator** [reply]
Sorry, I forgot to say thanks to you Frank! 
Thanks Frank so much for helping me out.
